core/src/test/java/org/apache/iceberg/TestSequenceNumberForV2Table.java

```diff
@@ -0,0 +1,299 @@
+        .collect(Collectors.toList()).get(0);
+    Assert.assertEquals(2, manifestFile.sequenceNumber());
+
+    manifestFile = writeManifest(FILE_C, FILE_D);
+    table.newFastAppend().appendManifest(manifestFile).commit();
+    Assert.assertEquals(3, table.currentSnapshot().sequenceNumber());
+
+    manifestFile = table.currentSnapshot().manifests().stream()
+        .filter(manifest -> manifest.snapshotId() == table.currentSnapshot().snapshotId())
+        .collect(Collectors.toList()).get(0);
+    Assert.assertEquals(3, manifestFile.sequenceNumber());
+
+    for (ManifestEntry entry : ManifestFiles.read(manifestFile,
+        table.io(), table.ops().current().specsById()).entries()) {
+      if (entry.file().path().equals(FILE_C.path()) || entry.file().path().equals(FILE_D.path())) {
+        Assert.assertEquals(3, entry.sequenceNumber().longValue());
+      }
+    }
+  }
+
+  @Test
+  public void testSequenceNumberForMergeAppend() throws IOException {
+    table.updateProperties()
+        .set(TableProperties.MANIFEST_MIN_MERGE_COUNT, "1")
+        .commit();
+    table.newAppend().appendFile(FILE_A).commit();
+    Assert.assertEquals(1, table.currentSnapshot().sequenceNumber());
+
+    table.newAppend().appendFile(FILE_B).commit();
+    Assert.assertEquals(2, table.currentSnapshot().sequenceNumber());
+
+    ManifestFile manifestFile = writeManifest(FILE_C, FILE_D);
+    table.newAppend().appendManifest(manifestFile).commit();
+    Assert.assertEquals(3, table.currentSnapshot().sequenceNumber());
+
+    manifestFile = table.currentSnapshot().manifests().get(0);
+
+    Assert.assertEquals("the sequence number of manifest should be 3", 3, manifestFile.sequenceNumber());
+
+    for (ManifestEntry entry : ManifestFiles.read(manifestFile,
+        table.io(), table.ops().current().specsById()).entries()) {
+      if (entry.file().path().equals(FILE_A.path())) {
+        Assert.assertEquals("the sequence number of data file should be 1", 1, entry.sequenceNumber().longValue());
+      }
+
+      if (entry.file().path().equals(FILE_B.path())) {
+        Assert.assertEquals("the sequence number of data file should be 2", 2, entry.sequenceNumber().longValue());
+      }
+
+      if (entry.file().path().equals(FILE_C.path()) || entry.file().path().equals(FILE_D.path())) {
+        Assert.assertEquals("the sequence number of data file should be 3", 3, entry.sequenceNumber().longValue());
+      }
+    }
+  }
+
+  @Test
+  public void testSequenceNumberForRewrite() {
+    table.newFastAppend().appendFile(FILE_A).commit();
+    table.newFastAppend().appendFile(FILE_B).commit();
+    Assert.assertEquals(2, table.currentSnapshot().sequenceNumber());
+
+    table.rewriteManifests().clusterBy(file -> "").commit();
+    Assert.assertEquals("the sequence number of snapshot should be 3",
+        3, table.currentSnapshot().sequenceNumber());
+
+    ManifestFile newManifest = table.currentSnapshot().manifests().get(0);
```

Comment:
This relies on ordering in the manifest list. Can you avoid assumptions like this?

Reply:
Sure, will use snapshot id to filter.